lib/Arch/Sleigh/MIPSArch.cpp

```diff
@@ -0,0 +1,379 @@
+
+llvm::Value *SleighMIPSDecoder::LiftPcFromCurrPc(llvm::IRBuilder<> &bldr,
+                                                llvm::Value *curr_pc,
+                                                size_t curr_insn_size,
+                                                const DecodingContext &) const {
+  return bldr.CreateAdd(curr_pc, llvm::ConstantInt::get(curr_pc->getType(), 4));
+}
+
+void SleighMIPSDecoder::InitializeSleighContext(
+    uint64_t addr, remill::sleigh::SingleInstructionSleighContext &ctxt,
+    const ContextValues &values) const {
+    //sleigh::SetContextRegisterValueInSleigh(
+    //  addr, std::string("ZERO").c_str(), "zero", 0, ctxt, values);
+}
+
+class SleighMIPSArch : public ArchBase {
+ public:
+  SleighMIPSArch(llvm::LLVMContext *context_, OSName os_name_,
+                ArchName arch_name_)
+      : ArchBase(context_, os_name_, arch_name_),
+        decoder(*this) {}
+  virtual ~SleighMIPSArch() = default;
+
+  DecodingContext CreateInitialContext(void) const override {
+    return DecodingContext();
+  }
+
+  std::string_view StackPointerRegisterName(void) const override {
+    return "SP";
+  }
+
+  std::string_view ProgramCounterRegisterName(void) const override {
+    return "PC";
+  }
+
+  OperandLifter::OpLifterPtr
+  DefaultLifter(const remill::IntrinsicTable &intrinsics) const override {
+    return decoder.GetOpLifter();
+  }
+
+  bool DecodeInstruction(uint64_t address, std::string_view instr_bytes,
+                         Instruction &inst,
+                         DecodingContext context) const override {
+    inst.pc = address;
+    inst.next_pc = address + instr_bytes.size();  // Default fall-through.
+    inst.branch_taken_pc = 0;
+    inst.branch_not_taken_pc = 0;
+    inst.has_branch_taken_delay_slot = false;
+    inst.has_branch_not_taken_delay_slot = false;
+    inst.arch_name = arch_name;
+    inst.sub_arch_name = arch_name;
+    inst.branch_taken_arch_name = arch_name;
+    inst.arch = this;
+    inst.category = Instruction::kCategoryInvalid;
+    inst.operands.clear();
+    inst.flows = Instruction::InvalidInsn();
+
+    context.UpdateContextReg(std::string("ZERO"), 0); // What to do here?
+
+    return this->decoder.DecodeInstruction(address, instr_bytes, inst, context);
+  }
+
+  uint64_t MinInstructionAlign(const DecodingContext &) const override {
+    return 4;
+  }
+
+  uint64_t MinInstructionSize(const DecodingContext &) const override {
+    return 4;
+  }
+
+  uint64_t MaxInstructionSize(const DecodingContext &,
+                              bool permit_fuse_idioms) const {
+    return permit_fuse_idioms ? 8 : 4;  // To handle delayslots, apparently
+  }
+
+  // Returns `true` if we should lift the semantics of `next_inst` as a delay
+  // slot of `inst`. The `branch_taken_path` tells us whether we are in the
+  // context of the taken path of a branch or the not-taken path of a branch.
+  bool NextInstructionIsDelayed(const Instruction &inst,
```

Comment:
Just copy paste junk as discussed, will be removed

Reply:
I checked, MayHaveDelaySlot is called by TraceLifter, so should NextInstructionIsDelayed (but i never see that logged), probably because has_branch_taken_delay_slot or so never get set?
Weird that delay slots work nontheless, i guess this isnt required for sleigh